sql-plugin/src/main/java/org/apache/spark/sql/catalyst/CudfUnsafeRow.java

```diff
@@ -0,0 +1,416 @@
+    int offset = 0;
+    for (Attribute attr : attributes) {
+      int length = GpuColumnVector.getRapidsType(attr.dataType()).getSizeInBytes();
+      offset = alignOffset(offset, length);
+      offset += length;
+    }
+    int bitSetWidthInBytes = calculateBitSetWidthInBytes(attributes.length);
+    // Each row is 64-bit aligned
+    return alignOffset(offset + bitSetWidthInBytes, 8);
+  }
+
+  //////////////////////////////////////////////////////////////////////////////
+  // Private fields and methods
+  //////////////////////////////////////////////////////////////////////////////
+
+  /**
+   * Address of where the row is stored in off heap memory.
+   */
+  private long address;
+
+  /**
+   * For each column the starting location to read from. The index to the is the position in
+   * the row bytes, not the user faceing ordinal.
+   */
+  private int[] startOffsets;
+
+  /**
+   * At what point validity data starts.
+   */
+  private int fixedWidthSizeInBytes;
+
+  /**
+   * The size of this row's backing data, in bytes.
+   */
+  private int sizeInBytes;
+
+  /**
+   * A mapping from the user facing ordinal to the index in the underlying row.
+   */
+  private int[] remapping;
+
+  /**
+   * Get the address of what a field is stored.
+   * @param i the index of the field in the row, not the user facing ordinal.
+   * @return the address of the field.
+   */
+  private long getFieldAddress(int i) {
+    return address + startOffsets[i];
+  }
+
+  /**
+   * Verify that index is valid for this row.
+   * @param index in this case the index can be either the user facing ordinal or the index into the
+   *              row.
+   */
+  private void assertIndexIsValid(int index) {
+    assert index >= 0 : "index (" + index + ") should >= 0";
+    assert index < startOffsets.length : "index (" + index + ") should < " + startOffsets.length;
+  }
+
+  //////////////////////////////////////////////////////////////////////////////
+  // Public methods
+  //////////////////////////////////////////////////////////////////////////////
+
+  /**
+   * Construct a new Row. The resulting row won't be usable until `pointTo()` has been called,
+   * since the value returned by this constructor is equivalent to a null pointer.
+   *
+   * @param attributes the schema of what this will hold.  This is the schema of the underlying
+   *                    row, so if columns were re-ordered it is the attributes of the reordered
+   *                    data.
+   * @param remapping a mapping from the user requested column to the underlying column in the
+   *                  backing row.
+   */
+  public CudfUnsafeRow(Attribute[] attributes, int[] remapping) {
+    int offset = 0;
+    startOffsets = new int[attributes.length];
+    for (int i = 0; i < attributes.length; i++) {
+      Attribute attr = attributes[i];
+      int length = GpuColumnVector.getRapidsType(attr.dataType()).getSizeInBytes();
```

Comment:
Worth asserting the length > 0 in case somehow this gets called for variable-width types?